atomic_reactor/utils/imageutil.py

```diff
@@ -92,33 +100,76 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
         session = util.RegistrySession.create_from_config(self._conf, registry)
         return util.RegistryClient(session)
 
+    def extract_file_from_image(self, image: str, src_path: str, dst_path: str) -> None:
+        """Extract file or directory from image at src_path to dst_path
 
-# OSBS2 TBD
-
-
-def get_image_history(image):
-    # get image history with skopeo / registry api
-    return []
-
-
-def inspect_built_image():
-    # get output image final/arch specific from somewhere
-    # and call get_inspect_for_image
-    return {}
-
-
-def remove_image(image, force=False):
-    # self.tasker.remove_image(image, force=force)
-    # most likely won't be needed at all
-    return {}
-
+        :param image: str, image pullspec from which to extract
+        :param src_path: str, path inside the image that points to file or directory
+                             that will be extracted
+        :param dst_path: str, path where to export file/dir
+        """
+        cmd = ['oc', 'image', 'extract', f'{image}', '--path', f'{src_path}:{dst_path}']
 
-def tag_image(image, new_image):
-    # self.tasker.tag_image(image, new_image)
-    return True
+        try:
+            retries.run_cmd(cmd)
+        except subprocess.CalledProcessError as e:
+            logger.error("Image file extraction failed\n%s", e.output)
+            raise
 
+    def download_image_archive_tarball(self, image: str, path: str) -> None:
+        """Downloads image archive tarball to path.
 
-def get_image(image):
-    # self.tasker.get_image(image)
-    # use skopeo copy
-    return {}
+        :param image: str, image pullspec to download
+        :param path: str, path including the filename of the tarball
+        """
+        cmd = ['skopeo', 'copy', f'docker://{image}', f'docker-archive://{path}']
+        try:
+            retries.run_cmd(cmd)
+        except subprocess.CalledProcessError as e:
+            logger.error("Image archive download failed:\n%s", e.output)
+            raise
+
+    def get_uncompressed_image_layer_sizes(self, path: str) -> List[Dict[str, Union[str, int]]]:
+        """Returns data about the uncompressed image layer sizes
+
+        :param path: str path to a image archive tarball
+        :return: List[Dict[str, object]], where key is layer id and int is the size
+                                 in bytes
+        """
+        with tarfile.open(path) as tar:
+            manifest_file = tar.extractfile('manifest.json')
+            if not manifest_file:
+                raise ValueError(f'manifest.json from {path} is not a regular file')
+            manifest = json.load(manifest_file)
+            layers = manifest[0]['Layers']
+            layer_sizes = []
+            for layer in layers:
+                layer_id = f'sha256:{layer.replace(".tar", "")}'
```

Comment:
Based on the above-mentioned spec, it seems that the file names of layers should not be relied on. The fact that they match the diff_id seems to be a skopeo implementation detail.

Instead, we could get the diff IDs from `rootfs.diff_ids` in the config file (also present in the docker archive and referenced in manifest.json) and rely on the fact that the `rootfs.diff_ids` array and the `Layers` array are in the same order.

Reply:
updated